Add TARDIS locator functionality to sonic scanning mode

Mansarde opened the pull request:
## About the PR
This PR adds a TARDIS locator ability to the sonic scanning mode.
When in scanning mode, interacting (right-clicking) on any Zeiton block will display the linked TARDIS' coordinates.

The Zeiton block requirement is to balance it a bit. The in-world idea is:
The sonic needs to interact with some Zeiton to make it resonate at the linked TARDIS' frequency, which sends a pulse through space-time that gets reflected by the linked TARDIS, kind of like space-time sonar.

Not sure if that is a satisfying explanation, but if you want other blocks as a requirement or don't want the block limitation at all, let me know. :)

## Why / Balance
It provides yet another alternative to get back to one's TARDIS, next to the Stattenheim Remote and the Hazandra.
The former gets the TARDIS to the player, if there's enough fuel.
The latter teleports the player into the interior, which is quite powerful, but requires preparation (i.e. linking the Hazandra before losing the TARDIS).
The locator functionality of this PR provides a last ditch effort to get to one's TARDIS, on foot if necessary.


## Technical details
During a scan of the sonic I check if the targeted block is one of the Zeiton blocks.
If so, I retrieve the current coordinates of the TARDIS that the sonic is linked to and display them to the user.

## Requirements
- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://amblelabs.github.io/ait-wiki/guidelines).
- [X] It does not require an ingame showcase.

**Changelog**
:cl:
- add: TARDIS locator functionality to sonic scanning mode

Saturnorsomthing commented:
sorta like NTM ig

Mansarde commented:
> just an idea but maby you could sonic a bell to make some cloister sound resonate

Hmm, that would also be good idea.
Gonna walk the dog, but will play around with that once I get back.

duzos commented on src/main/java/dev/amble/ait/core/item/sonic/ScanningSonicMode.java:
```diff
@@ -99,6 +102,27 @@ public boolean scanBlocks(ItemStack stack, World world, PlayerEntity user, Block
 
         String blastRes = String.format("%.2f", block.getBlastResistance());
 
+        Block[] validLocatorBlocks = {
```
Use block tags

Mansarde replied:
Gotcha, will do.

duzos commented on src/main/java/dev/amble/ait/core/item/sonic/ScanningSonicMode.java:
```diff
@@ -99,6 +102,27 @@ public boolean scanBlocks(ItemStack stack, World world, PlayerEntity user, Block
+
+        if (Arrays.asList(validLocatorBlocks).contains(block))
+        {
+            Tardis tardis = SonicItem.getTardisStatic(world, stack);
+            int x = tardis.travel().position().getPos().getX();
+            int y = tardis.travel().position().getPos().getY();
+            int z = tardis.travel().position().getPos().getZ();
+            Text message = Text.translatable("item.sonic.scanning.locator_message").append(String.format("%s %s %s", x, y, z));
```
```suggestion
            Tardis tardis = SonicItem.getTardisStatic(world, stack);
            BlockPos tPos = tardis.travel().position().getPos();
            Text message = Text.translatable("item.sonic.scanning.locator_message").append(String.format("%s %s %s", tPos.getX(), tPos.getY(), tPos.getZ()));
```

duzos replied:
I'd also recommend you check / send something related to dimensions

Mansarde commented:
Ok, I got almost everything done (block tags, shorter getpos and dimension name).
Only thing left is to implement Saturn's idea for the bell.
But I won't get that done tonight, will resume tomorrow and push everything once it's done.

Mansarde commented:
@duzos I've pushed three new commits that address your change requests.
I also added the bell as a valid block for the locator functionality.
For that I had to hook into the `UseBlockCallback` event in the `AITMod` class.
Hope I did everything correct with that, would be nice if you could take a look and check if everything is in order there.
Thx!

Mansarde commented on src/main/java/dev/amble/ait/AITMod.java:
```diff
@@ -255,6 +261,22 @@ public void onInitialize() {
             SetDoorParticleCommand.register(dispatcher, registryAccess);
         }));
 
+        UseBlockCallback.EVENT.register((player, world, hand, hitResult) -> {
```
Roger that, I will try to do it with a mixin instead,

Mansarde commented on src/main/java/dev/amble/ait/core/item/sonic/ScanningSonicMode.java:
```diff
@@ -99,6 +104,20 @@ public boolean scanBlocks(ItemStack stack, World world, PlayerEntity user, Block
+            Tardis tardis = SonicItem.getTardisStatic(world, stack);
+            BlockPos tPos = tardis.travel().position().getPos();
+            String dimensionText = MonitorUtil.truncateDimensionName(WorldUtil.worldText(world.getRegistryKey()).getString(), 20);
+            String locationText = String.format("%s\nCoordinates: %s %s %s", dimensionText, tPos.getX(), tPos.getY(), tPos.getZ());
```
Just for the coordinates, or for the dimension name as well?

Mansarde commented:
@DrTheodor I just pushed new commits that address your two points above.
I.e. I replaced the UseBlockCallback with a mixin for the bell block, and added the text "Coordinates" as a translatable.

duzos reviewed (dismissed):
LGTM

DrTheodor commented on src/main/java/dev/amble/ait/core/item/sonic/ScanningSonicMode.java:
```diff
@@ -99,6 +104,21 @@ public boolean scanBlocks(ItemStack stack, World world, PlayerEntity user, Block
         String blastRes = String.format("%.2f", block.getBlastResistance());
 
+        if (state.isIn(AITTags.Blocks.SONIC_CAN_LOCATE))
+        {
```
no new line should be here, we follow java style guidelines, not c#

Mansarde replied:
Ah I still had that in a new line because previously the if block was multi-line and I wanted to make it clear where the the interior body of the if begins.
But will keep the opening bracket on the previous line from now on (I prefer that anyways ^_^).
Thx!

DrTheodor commented on src/main/java/dev/amble/ait/core/item/sonic/ScanningSonicMode.java:
```diff
@@ -99,6 +104,21 @@ public boolean scanBlocks(ItemStack stack, World world, PlayerEntity user, Block
+            Tardis tardis = SonicItem.getTardisStatic(world, stack);
+            BlockPos tPos = tardis.travel().position().getPos();
+            String dimensionText = MonitorUtil.truncateDimensionName(WorldUtil.worldText(world.getRegistryKey()).getString(), 20);
+            String coordinatesText = Text.translatable("item.sonic.scanning.locator_message.coordinates").getString();
```
this isnt how its done
you can use %s inside translations and put the arguments after the translation key

Mansarde replied:
Roger that, will adjust that accordingly.

Mansarde replied:
@DrTheodor 
Ok, changed it so that the format specifiers are now in the translation itself.
I still put the newline as a separate `append("\n")` because I didn't think it'd be good to have it in the translation.
But let me know if you find that's better and I'll adjust it accordingly.

Mansarde commented:
For now please do not merge yet.
IT's ok for review regarding the new commits, but there's still some small changes I need to make to the mixin.

Mansarde commented:
Alright, it's ready for review again. :)
